### PR TITLE
fix: redirect stream when get last replace

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
When replace a DDL into multiple DDLs, eg:
```
1000-1010 --> 1000.0-1000.1, 1000.1-1010.0
```
When we redirect to 1000.1 in resharding, after get replace event, stream still at 1000. It will get origin event again

### What is changed and how it works?
redirect stream when get last replace event.

### Check List 

Tests 

 - Integration test

### CODE REVIEW

File: syncer/syncer.go [hunk 1]
```diff
@@ -1331,11 +1331,21 @@ func (s *Syncer) Run(ctx context.Context) (err error) {
...
-		// we calculate endLocation(currentLocation) for Rows/Query event here
+		// we calculate startLocation and endLocation(currentLocation) for Rows/Query event here
+		// set startLocation empty for other events to avoid misuse
+		startLocation = binlog.Location{}
 		switch e.Event.(type) {
 		case *replication.RowsEvent, *replication.QueryEvent:
+			startLocation = binlog.Location{
+				Position: mysql.Position{
+					Name: lastLocation.Position.Name,
+					Pos:  e.Header.LogPos - e.Header.EventSize,
```
Comment (lance6716) [inline syncer/syncer.go:1343]: can't use lastLocation?

**Comment by lance6716** [comment]
we might list these XXXLocations' meaning in comment, with some invarant properties attached to them (such as "at beginning of for-loop, events before XXXLocation event has been synced" )

**Comment by lance6716** [comment]
(please help cherry-pick this https://github.com/pingcap/dm/commit/6d1aede47719764e2eac05009255833c87f349c2 to this PR) 😂

File: syncer/syncer.go (hunk 1 above)
Comment (GMHDBJD) [reply syncer/syncer.go:1343]: More safe. 😂

**Comment by GMHDBJD** [comment]
> we might list these XXXLocations' meaning in comment, with some invarant properties attached to them (such as "at beginning of for-loop, events before XXXLocation event has been synced" )

I update the comment in the beginning. But most comments are still scattered in different places.

**Comment by lance6716** [comment]
advice some renaming:

```
currentLocation = s.checkpoint.GlobalPoint() // also init to global checkpoint
```
rename it to `nextLocation`, aware there's a `nextLocation`around line 1212, may need rename it to `...2`

--

```
shardingReSync.currLocation
```
ditto

--

```
		s.currentLocationMu.Lock()
		s.currentLocationMu.currentLocation = currentLocation.Clone()
		s.currentLocationMu.Unlock()
```
this is indeed the start location of "current" event, and we use `startLocation` representing "start location of current event". So how about a) renaming it to `s.startLocationMu` or b) renaming `startLocation` to `currentLocation`

And here we forward the "current", how about change "next" location also. So later code could reference a clear variable

--

```
e, err = s.getEvent(tctx, &currentLocation)
```
maybe `s.getEvent(tctx, &s.startLocationMu.startLocation)` (depend on above renaming) to make it clear?

--

```
startLocation = binlog.Location{}
...
			startLocation = binlog.Location{
				Position: mysql.Position{
					Name: lastLocation.Position.Name,
					Pos:  e.Header.LogPos - e.Header.EventSize, // <- e is current event
```
in fact it's `prevLocation` right?

--

also align them with `eventContext`, `ShardingReSync` structure

**Comment by lance6716** [comment]
LGTM

**Pull request merged**